kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java

```diff
@@ -778,6 +777,50 @@ void testSendRequestTimeout() {
         assertThrows(ExecutionException.class, q::get);
     }
 
+    @Test
+    void sendRequestChainedActionsRunOnNettyEventLoop() {
+        var eventLoopThreadFuture = new CompletableFuture<Thread>();
+
+        var sendRequestBody = new FetchRequestData();
+        ApiVersionsRequestFilter filter = (apiVersion, header, request, context) -> context.sendRequest((short) 3, sendRequestBody)
+                .thenApply(u1 -> {
+                    assertThat(Thread.currentThread())
+                            .describedAs("first chained action must run on event loop")
+                            .isEqualTo(eventLoopThreadFuture.getNow(null));
+
+                    return null;
+                }).thenApplyAsync(u2 -> {
+                    assertThat(Thread.currentThread())
+                            .describedAs("second chained action must also run on event loop")
+                            .isEqualTo(eventLoopThreadFuture.getNow(null));
+                    return null;
+                }).thenCompose(u3 -> context.forwardRequest(header, request));
+
+        buildChannel(filter);
+
+        // capture the thread used by the embedded channel
+        channel.eventLoop().submit(() -> eventLoopThreadFuture.complete(Thread.currentThread()));
```

Comment:
Note to self: the embedded channel's event loop is the main thread.
It surprised me to learn that  `io.netty.channel.embedded.EmbeddedEventLoop#inEventLoop()` is hardcoded to return true.  So it lies if called from another thread.  In my first iteration of this test I had in inEventLoop assertion passing even when called from the forkJoin pool thread.